### PR TITLE
Artist Registry

### PR DESCRIPTION
**Pull request by JamesDalboth**
### Your checklist for this pull request
🚨Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a feature/bugfix branch** (right side). Don't request from master!
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.

### Description
I have implemented an Artist Registry if you will. This allows us to grab a list of approved artists instead of manually typing their name, on the front end. After this I will release a new MR to remove questions like artist nationality and birth date since this can be stored with the artist on the backend

### Checklist
* [x] Have you done your changes on a separate branch. Branches MUST have descriptive names that start with either the `fix/` or `feature/` prefixes. Good examples are: `fix/signin-loop` or `feature/pr-templates`.
* [x] Have you got descriptive commit messages that would make sense if prefixed with "This commit will ..."
* [x] Will you squash when you merge this PR?

#### New Features

1. [x] Does your submission have a high coverage of your feature?
2. [x] Have you lint your code locally prior to submission?

#### Bugs

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests which cover this bug?
* [x] Have you successfully ran tests with your changes locally?

💔Thank you!

### CODE REVIEW

**Comment by felination** [comment]
Need to run a lint -fix on the front end

File: app/src/components/Register.tsx [hunk 1]
```diff
@@ -62,6 +68,45 @@ class Register extends React.Component<RegisterProps, RegisterState> {
...
+                name: info[0],
+                wallet: info[1],
+              };
+            }));
+        }
+
+        return Promise.all(artists);
+      })
+      .then((artists: Artist[]) => this.setState({ artists: artists }))
+      .catch((err: any) => console.log(err));
```
Comment (DJRHails) [inline app/src/components/Register.tsx]: Separate out into async function and then call that inside here?

File: app/src/components/Register.tsx [hunk 2]
```diff
@@ -145,6 +190,20 @@ class Register extends React.Component<RegisterProps, RegisterState> {
...
+  getOptions = (): any => {
+    if (!this.state.artists) {
+      return [];
+    }
+
+    const options = [];
+
+    for (const artist of this.state.artists) {
+      options.push(<option>{artist.name}</option>);
+    }
```
Comment (DJRHails) [inline app/src/components/Register.tsx]: Use a map? Might be a little bit easier to comprehend

File: contracts/Artists.sol [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    _transferOwnership(owner);
+
+    // Pre populate with some artists
+    addArtist("Vincent Van Gogh", address(uint(keccak256(abi.encodePacked("Vincent Van Gogh")))));
+    addArtist("Pablo Picasso", address(uint(keccak256(abi.encodePacked("Pablo Picasso")))));
+    addArtist("Leonardo da Vinci", address(uint(keccak256(abi.encodePacked("Leonardo da Vinci")))));
+    addArtist("Claude Monet", address(uint(keccak256(abi.encodePacked("Claude Monet")))));
+    addArtist("Andy Warhol", address(uint(keccak256(abi.encodePacked("Andy Warhol")))));
+    addArtist("Salvador Dali", address(uint(keccak256(abi.encodePacked("Salvador Dali")))));
+    addArtist("Michelangelo", address(uint(keccak256(abi.encodePacked("Michelangelo")))));
```
Comment (DJRHails) [inline contracts/Artists.sol]: Move into a migration?

Comment (DJRHails) [reply contracts/Artists.sol]: Are these addresses just meant to be fake or is there some logic to these?

File: migrations/7_deploy_artists.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const Artists = artifacts.require('Artists');
+const ENSResolver = artifacts.require('ENSResolver');
+const FIFSRegistrar = artifacts.require('FIFSRegistrar');
+
+const newLabel = require('./helper/LoggedRegistration');
+
+module.exports = async (deployer, network, accounts) => {
+  await deployer.deploy(Artists, accounts[0], { from: accounts[0] });
```
Comment (DJRHails) [inline migrations/7_deploy_artists.js:8]: Do you want to do the governance trick with this? So governance controls artists?

File: test/Artists.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+const { expectRevert } = require('@openzeppelin/test-helpers');
+
+const Contract = artifacts.require('Artists');
```
Comment (DJRHails) [inline test/Artists.js]: Style normally has the contract name as Artists over Contract, but not a big one

**Review by DJRHails — COMMENTED** [review]
Looks good in principal, couple of small comments I had reading through.

File: app/src/components/Register.tsx (hunk 1 above)
Comment (DJRHails) [reply app/src/components/Register.tsx]: Also the same comment as below with the mapping trick.

File: contracts/Artists.sol (hunk 3 above)
Comment (JamesDalboth) [reply contracts/Artists.sol]: Good idea to move to migration, no I just needed a way to generate addresses

File: migrations/7_deploy_artists.js (hunk 4 above)
Comment (DJRHails) [reply migrations/7_deploy_artists.js:8]: Probably out of scope but there needs to be some mechanism to add (trello task at low priority?)

Comment (felination) [reply migrations/7_deploy_artists.js:8]: This is what I came here to say: what happens when the piece is by an artist not yet in our system

**Comment by felination** [comment]
Where is the example text for each card generated? We should note that and make a low priority trello task to have it be slightly less generic

**Comment by felination** [comment]
I presume this was done in Dan's PR earlier but the Artifact information / artist information tabs are unclear and confused me initially

**Comment by JamesDalboth** [comment]
For both comments above not relevant to my task but agree

File: migrations/7_deploy_artists.js (hunk 4 above)
Comment (JamesDalboth) [reply migrations/7_deploy_artists.js:8]: To governance and system to add, long term I agree but would rather spin off other tasks to reduce size of PR. I also don't think priority wise most important.

File: app/src/components/Register.tsx (hunk 1 above)
Comment (JamesDalboth) [reply app/src/components/Register.tsx]: I might be missing the obvious but where do you want me to apply the mapping trick in this case? I don't really do any list to list transformations

Comment (DJRHails) [reply app/src/components/Register.tsx]: Turning from a stream of integers between 0 and total into an artist object { name, wallet }

File: contracts/Artists.sol (hunk 3 above)
Comment (DJRHails) [reply contracts/Artists.sol]: I'd recommend putting a zero address in for the moment then so we know they're not set

File: app/src/components/Register.tsx [hunk 6]
```diff
@@ -60,7 +66,53 @@ class Register extends React.Component<RegisterProps, RegisterState> {
...
+  getArtistInfo = (): Promise<Artist[]> => {
+    const Artists = this.props.drizzle.contracts.Artists;
+
+    return Artists.methods.getArtistsTotal()
+      .call()
+      .then((total: number) => {
+        const artists = [];
+
+        let id = 0;
+        while (id < total) {
```
Comment (DJRHails) [inline app/src/components/Register.tsx:103]: See previous comment about how to map this one

File: app/src/components/Register.tsx [hunk 7]
```diff
@@ -145,6 +197,14 @@ class Register extends React.Component<RegisterProps, RegisterState> {
     this.setState(stateUpdate);
   };
 
+  getOptions = (): any => {
+    if (!this.state.artists) {
+      return [];
+    }
+
+    return this.state.artists.map((artist: Artist) => <option key={artist.wallet}>{artist.name}</option>);
```
Comment (DJRHails) [inline app/src/components/Register.tsx:205]: Love it

File: app/src/components/Register.tsx [hunk 8]
```diff
@@ -145,6 +197,14 @@ class Register extends React.Component<RegisterProps, RegisterState> {
     this.setState(stateUpdate);
   };
 
+  getOptions = (): any => {
```
Comment (DJRHails) [inline app/src/components/Register.tsx]: Any?

**Review by DJRHails — CHANGES_REQUESTED** [review]

**Comment by DJRHails** [comment]
Also if you trello task the other stuff we've mentioned just so we can keep track of it

File: app/src/components/Register.tsx (hunk 1 above)
Comment (JamesDalboth) [reply app/src/components/Register.tsx]: I would have to create a loop to create that list of integers between 0 to total. and then do a map. Not sure that would look cleaner

File: app/src/components/Register.tsx (hunk 8 above)
Comment (JamesDalboth) [reply app/src/components/Register.tsx]: Ah yes I can't work out how to type a list of <option>, what type is <option>?

**Pull request merged**